### ISSUE TITLE
No alternative to unfocus

### ISSUE THREAD

**Issue by rnrand** [coq/coq#10688]
#### Description of the problem

Coq 8.8 deprecated the Focus and Unfocus commands, recommending that we use n:{ } in place of focus but recommending nothing to replace Unfocus. 

Unfocus is important though. If I have a bunch of goals that contain evars, there's no guarantee of an order in which I can solve the goals (assuming I don't know what all the evars are beforehand). 

I can, of course, write my proof fragment using `n:{ a. b. c.` and then rewrite it using `n: a. n: b. n:c`, but that makes my proofs unreadable. 

See https://github.com/inQWIRE/QWIRE/blob/55399b856cddfac0d932da98fa26ffd9d693d0d4/TypeChecking.v#L377-L408 for a lemma where Unfocus is important.

#### Coq Version

8.8+

**Comment by Zimmi48** [comment]
Without executing it, it is still a bit hard to understand the goal flow in your proof:

```coq
    + destruct_merges; subst.
      eapply @types_pair.
      4: eauto.
      Focus 3.
        econstructor.
        3: eauto.
        3: eauto.
        2: monoid.
        validate.
      Unfocus.
      2: monoid.
      subst; validate.
      subst; validate.
```

Can I consider that the proof is structured as if the following option was set?

```
Set Default Goal Selector "!".
```

(Cf. https://coq.inria.fr/refman/proof-engine/tactics.html#coq:opt.default-goal-selector)

Could it be refactored into the following?

```coq
    + destruct_merges; subst.
      eapply @types_pair.
      4:{ eauto. }
      Focus 3.
        econstructor. (* generates 5 sub-goals *)
        3:{ eauto. }
        3:{ eauto. }
        2:{ monoid. }
        { validate. }
        (* a single goal left over *)
      Unfocus.
      2:{ monoid. }
      * subst; validate.
      * subst; validate.
```

If I understood correctly, let me reproduce the same proof flow with dummy goals:

```coq
Definition foo (n : nat) := True.

Goal foo 1 /\ foo 2 /\ foo 3.
split.
Focus 2.
  split.
  2: easy.
Unfocus.
* easy.
* easy.
Qed.
```

And now, let me do the same without `Unfocus` (requires Coq 8.10 for the curly braces to support named goal selectors):

```coq
Ltac name_goal name := refine ?[name].

Goal foo 1 /\ foo 2 /\ foo 3.
split.
2:{
  split.
  2: easy.
  name_goal foo_2; shelve.
}
easy.
[foo_2]: { easy. }
Qed.
```

Said otherwise, you can always have goals escape the curly braces thanks to `shelve`, and the named goals are a nice way of unshelving later without losing the track of what goal comes from where.

---

BTW, I wanted to ask earlier on Coq-Club whether you had considered submitting QWIRE for inclusion in Coq's CI. Cf. https://github.com/coq/coq/blob/master/dev/ci/README-users.md.

**Comment by rnrand** [comment]
That's pretty much correct, save that the validate before `Unfocus` doesn't solve a subgoal. But that's fine, I don't need it there.

So here's the subproof with `Set Default Goal Selector "!".`


```
    + destruct_merges; subst.
      eapply @types_pair.
      4:{ eauto. }
      Focus 3.
        econstructor. (* generates 5 sub-goals *)
        3:{ eauto. }
        3:{ eauto. }
        2:{ monoid. }
        (* a single goal left over *)
      Unfocus.
      2:{ monoid. }
      * subst; validate.
      * subst; validate.
```

And here's the proof following your suggestion:

```
    + destruct_merges; subst.
      eapply @types_pair.
      4:{ eauto. }
      3:{
        econstructor. (* generates 5 sub-goals *)
        3:{ eauto. }
        3:{ eauto. }
        2:{ monoid. }
        (* a single goal left over *)
        name_goal V1. shelve.
        }
      2:{ monoid. }
      1:{ subst; validate. }
      [V1]: { easy. }
```

(I'm supposing that this works. I actually can't seem to get the master branch or 8.10beta to build, so I'm on 8.9.)

If I don't want to (or can't) use named goals I can do the following:

```
    shelve. }
    Unshelve.
      2:{ monoid. }
      * subst; validate.
      * subst; validate.
```

Both solutions seem pretty undesirable, though. One thing you can't see from the proof above is that once I call `Unshelve` I now have 4 goals: the three I had before plus an evar that is solved en passant by 2:{ monoid }. (Adding an `all: shelve_unifiable.` helps here.) In principle, I could have any number of nonsensical goals.

Using either approach, it's my responsibility to solve V1 in the right place. I can easily forget that V1 exists, and I won't be reminded until I attempt Qed -- this is why I don't like the shelf in general. 

I'd much prefer a `!}` (or the like) that closes the scope while leaving the goal "in place". (And then `Unfocus` could tell me what to use instead, which is important.)

**Comment by rnrand** [comment]
> BTW, I wanted to ask earlier on Coq-Club whether you had considered submitting QWIRE for inclusion in Coq's CI. Cf. https://github.com/coq/coq/blob/master/dev/ci/README-users.md.

I've thought about it. One of the biggest obstacles would be the requirement:

> You maintain a reasonable build time for your development, or you provide a "lite" target that we can use.

I could use more guidance on what constitutes a "reasonable build time". 

Right now QWIRE has `make` (formerly `make core`) and `make all` options, the former of which only compiles the core of QWIRE. Even that takes a while, though. I'd also consider setting up CI for SQIRE (which imports libraries from QWIRE) but it has similar issues.

**Comment by Zimmi48** [comment]
FWIW our CI currently includes some heavy projects like bedrock2 (65 minutes build), fiat-crypto (95 minutes build), lambda-rust with its dependencies (60 minutes build), math-comp (95 minutes build), and VST (75 minutes build).

> I'd much prefer a `!}` (or the like) that closes the scope while leaving the goal "in place". (And then `Unfocus` could tell me what to use instead, which is important.)

Sounds like a reasonable feature request. FTR, I didn't implement such thing in the first place because I searched for uses of `Unfocus` on GitHub, and through many pages of results, I could only find a pattern where it was easily replaced by `n: { }`. But there is no pressing need to remove `Unshelve` so let's not do it until `!}` (or whatever syntax it ends up with) gets implemented.

**Comment by ejgallego** [comment]
> I actually can't seem to get the master branch or 8.10beta to build

Likely your OCaml version is too old.

> I could use more guidance on what constitutes a "reasonable build time".

The current guideline is less than 1 hour in our 2-core CI machines, we are flexible on that tho.

**Comment by Zimmi48** [comment]
What do you think the ordering of the goals after `!}` should be? Do you think that the ordering of `Unfocus` (inserts all the remaining goals where we initially focused) should be kept, or would you think that a simpler semantics (all the remaining goals are inserted at the beginning or at the end) would be OK? The second solution could allow a slight simplification of the way the proof stack is managed.

**Comment by rnrand** [comment]
I do think the ordering of `Unfocus` should be kept, not because it's particularly intuitive in the case where new subgoals are generated (I'd honestly prefer a clearer tree-like structure) but because `2:{ split. !}` should be identical to `2: split.` Otherwise, it's likely to be confusing.

**Comment by Zimmi48** [comment]
Yes, that seems reasonable.

**Comment by Zimmi48** [comment]
@rnrand 
> I'd honestly prefer a clearer tree-like structure

I was pondering over this and wondering what exactly you meant by this. Do you think the right choice was made with the behavior of goal selectors (that goals generated by `2: foo.` are inserted between the first and the third goals) or not? This choice is rather intuitive when a tactic doesn't change the number of goals. However, the less intuitive part is that when multiple goals are selected, all the sub-goals end up in the position of the first selected goal. If we were to consider instead that the subgoals generated by `2: foo.` are inserted before all the remaining goals, then this would remove the issue with multiple goal selectors. This would also present the advantage of making `2: {` an equivalent of `2: idtac. {`. If compatibility is the concern, we could provide a flag to choose which behavior is preferred. But for `!}`, since that would be a new feature, we could choose whichever behavior we think is the most natural (and say that `2: foo.` is equivalent to `2: { foo !}` with the right flag turned on).

**Comment by rnrand** [comment]
What I mean is that `destruct a; destruct b; destruct c` fundamentally produces a tree. Assuming `a`, `b` and `c` are booleans, I should focus on the goal where `a=false, b=true, c=false` by means of `2. 1. 2.` (using an ad-hoc syntax). 

Indeed, if I write it in the following manner, I get exactly that structure:

```
destruct a.
- destruct b.
   + destruct c.
     * ...
     * ...
   + destruct c.
     * ...
     * ...
- destruct b.
   + destruct c.
     * ...
     * ...
   + destruct c.
     * ...
     * ...
```
and this makes the structure of the proof dramatically clearer. (Though sequencing destructs or failing to focus immediately after doing case analysis will ruin things.)

But given that `destruct a; destruct b; destruct c` gives a list of goals rather than a tree, I'd prefer behavior that is consistent with the `2:` notation. That is, slot the new goals in between 2 and 3, rather than at the end.

**Comment by SkySkimmer** [comment]
IIUC the idea is to provide new syntax involving a `}`  for Unfocus?

**Comment by Zimmi48** [comment]
Yes. The proposed syntax was `!}`, but it could be something else.
